File: Battelle.EPA.WideAreaDecon.Client/src/implementations/parameter/distribution/FrequencyValue.ts
import { JsonProperty, Serializable } from 'typescript-json-serializer';
import ParameterType from '@/enums/parameter/parameterType';
import IParameter from '@/interfaces/parameter/IParameter';
import IUnivariateParameter from '@/interfaces/parameter/IUnivariateParameter';
import ParameterMetaData from '../ParameterMetaData';
import FrequencyValueType from '@/enums/parameter/frequencyValueType';

@Serializable()
export default class FrequencyValue implements IUnivariateParameter {
  @JsonProperty()
  readonly type: ParameterType = ParameterType.frequencyValue;

  @JsonProperty()
  metaData: ParameterMetaData;

  public get isSet(): boolean {
    return this.value !== null && this.value !== undefined;
  }

  @JsonProperty()
  public value: FrequencyValueType;

  public get textValue(): FrequencyValueType {
    return this.value;
  }

  constructor(metaData = new ParameterMetaData(), value: FrequencyValueType) {
    this.value = value ?? null;
    this.metaData = metaData;
  }

  isEquivalent(other: IParameter): boolean {
    return this.compareValues(other as FrequencyValue);
  }

  compareValues(other?: FrequencyValue): boolean {
    return other ? this.type === other.type && this.value === other.value : false;
  }
}

File: Battelle.EPA.WideAreaDecon.InterfaceData/Models/Parameter/FrequencyValue.cs
using System;
using Battelle.EPA.WideAreaDecon.InterfaceData.Enumeration.Parameter;
using Battelle.EPA.WideAreaDecon.InterfaceData.Interfaces.Parameter;
using Battelle.EPA.WideAreaDecon.InterfaceData.Utility.Attributes;
using Battelle.EPA.WideAreaDecon.InterfaceData.Utility.Extensions;
using Battelle.EPA.WideAreaDecon.InterfaceData.Utility.Helpers;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using NPOI.SS.UserModel;
using Stats = Battelle.RiskAssessment.Common.Statistics;

namespace Battelle.EPA.WideAreaDecon.InterfaceData.Models.Parameter
{
    /// <summary>
    /// Implementation of the frequency value handling from data sheet
    /// </summary>
    public class FrequencyValue : IParameter
    {
        [JsonConverter(typeof(StringEnumConverter))]
        public ParameterType Type => ParameterType.Frequency;

        public ParameterMetaData MetaData { get; set; }

        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
        [ExcelProperty(ParameterLocationHelper.Parameter1)]
        public FrequencyValueType Value { get; set; }

        public static FrequencyValue FromExcel(ParameterMetaData metaData, IRow row)
        {
            // Parse the frequency value string as the enum type FrequencyValueType and throw an exception
            // if the string cannot be parsed into the correct type
            if (!Enum.TryParse(typeof(FrequencyValue).GetCellValue(nameof(Value), row), out FrequencyValueType value))
            {
                throw new ApplicationException($"{typeof(FrequencyValue).GetCellValue(nameof(Value), row)} is not a valid frequency value. Please select either 'Never', 'Once', or 'Multiple'.");
            }

            // Return the frequency value object
            return new FrequencyValue()
            {
                Value = value,
                MetaData = metaData
            };
        }

        public Stats.IDistribution CreateDistribution()
        {
            throw new NotImplementedException();
        }

        public string GetTextValue()
        {
            throw new NotImplementedException();
        }

        public FrequencyValueType GetFrequencyValue()
        {
            return Value;
        }
    }
}
